File: Phase5/Search/InvertedIndex.cs
using System.Collections.Generic;

namespace Phase5
{
    public class InvertedIndex : IInvertedIndex
    {
        private readonly Dictionary<string, HashSet<string>> _wordsMap;

        public InvertedIndex()
        {
            _wordsMap = new Dictionary<string, HashSet<string>>();
        }

        public void BuildInvertedIndex(Dictionary<string, string> docsMap, ITokenizer tokenizer)
        {
            foreach (var (docId, docContent) in docsMap)
            {
                var docWords = tokenizer.Tokenize(docContent);
                AddDoc(docWords, docId);
            }
        }
        public HashSet<string> GetWordDocs(string word)
        {
            return _wordsMap.GetValueOrDefault(word, new HashSet<string>());
        }
        private void AddDoc(HashSet<string> docWords, string docId)
        {
            foreach (var word in docWords)
            {
                if (!_wordsMap.ContainsKey(word))
                    _wordsMap.Add(word, new HashSet<string>());
                _wordsMap[word].Add(docId);
            }
        }
        
    }
}

File: Phase1/src/InvertedIndex.java
package src;

import java.util.HashMap;
import java.util.HashSet;

public class InvertedIndex {
    private final HashMap<String, HashSet<Integer>> wordsMap;

    public InvertedIndex() {
        wordsMap = new HashMap<>();
    }

    public void addDoc(HashSet<String> docWords, int docId) {
        for (String word : docWords) {
            if (word.length() <= 1)
                continue;
            addWord(word);
            addDocID(word, docId);
        }
    }

    private void addDocID(String word, int ID) {
        wordsMap.get(word).add(ID);
    }

    private void addWord(String word) {
        if (!wordsMap.containsKey(word))
            wordsMap.put(word, new HashSet<>());
    }

    public HashSet<Integer> getWordDocs(String word) {
        return wordsMap.getOrDefault(word, new HashSet<Integer>());
    }

    public HashMap<String, HashSet<Integer>> getWordsMap() {
        return wordsMap;
    }
}
